test_mapbased_param/Cl_estimation.py

```diff
@@ -0,0 +1,375 @@
+        instrument = {'frequencies':np.array(self.config['frequencies'])}
+        ind = 0
+        noise_cov_ = np.zeros((len(instrument['frequencies']), 3, frequency_maps.shape[-1]))
+        for f in range(len(instrument['frequencies'])) : 
+            for i in range(3): 
+                noise_cov_[f,i,:] = noise_cov[ind,:]*1.0
+                ind += 1
+        # removing I from all maps
+        noise_cov_ = noise_cov_[:,1:,:]
+
+
+        nside_map = hp.get_nside(clean_map[0])
+        print('nside_map = ', nside_map)
+        
+        w=nmt.NmtWorkspace()
+        b = binning_definition(self.config['nside'], lmin=self.config['lmin'], lmax=self.config['lmax'],\
+                                         nlb=self.config['nlb'], custom_bins=self.config['custom_bins'])
+
+        print('building mask ... ')
+        mask =  hp.read_map(self.get_input('binary_mask_cut'))
+        obs_pix = np.where(mask!=0)[0]
+        
+
+        if self.config['mask_apo'] != '':
+            mask_apo = hp.read_map(self.config['mask_apo'], verbose=False, field=None, h=False)
+            mask_apo = hp.ud_grade(mask_apo, nside_out=self.config['nside'])
+        else:
+            if self.config['extra_apodization']:
+                mask_apo = nmt.mask_apodization(mask, self.config['aposize'], apotype=self.config['apotype'])
+            else: mask_apo = mask*1.0
+
+            if ((self.config['noise_option']!='white_noise') 
+                    and (self.config['noise_option']!='no_noise')):
+                        # and (not self.config['no_inh'])):
+                nh = hp.smoothing(nh, fwhm=1*np.pi/180.0, verbose=False) 
+                nh /= nh.max()
+                mask_apo *= nh
+
+        fsky_eff = np.mean(mask_apo)
+        print('fsky_eff = ', fsky_eff)
+        np.savetxt(self.get_output('fsky_eff'), [fsky_eff])
+
+        print('building ell_eff ... ')
+        ell_eff = b.get_effective_ells()
+
+        #Read power spectrum and provide function to generate simulated skies
+        cltt,clee,clbb,clte = hp.read_cl(self.config['Cls_fiducial'])[:,:4000]
+        mp_t_sim,mp_q_sim,mp_u_sim=hp.synfast([cltt,clee,clbb,clte], nside=nside_map, new=True, verbose=False)
+
+        def get_field(mp_q, mp_u, mask_apo, purify_e=False, purify_b=True) :
+            #This creates a spin-2 field with both pure E and B.
+            f2y=nmt.NmtField(mask_apo,[mp_q,mp_u],purify_e=purify_e,purify_b=purify_b)
+            return f2y
+
+        #We initialize two workspaces for the non-pure and pure fields:
+        # if ((self.config['noise_option']!='white_noise') and (self.config['noise_option']!='no_noise')):
+            # f2y0=get_field(mask_nh*mp_q_sim,mask_nh*mp_u_sim)
+        # else:
+        f2y0=get_field(mp_q_sim,mp_u_sim,mask_apo)
+        w.compute_coupling_matrix(f2y0,f2y0,b)
+
+        #This wraps up the two steps needed to compute the power spectrum
+        #once the workspace has been initialized
+        def compute_master(f_a,f_b,wsp) :
```

Comment:
you can also just write `w.decouple_cell(nmt.compute_coupled_cell(f_a, f_b))` every time if you don't want to have to define this here.